_posts/2024-01-22-Python-Memory-Management-and-Troubleshooting.md

```diff
@@ -0,0 +1,568 @@
+
+style computer fill:#fff,stroke:#333
+style memory fill:#86EFAC,stroke:#333
+style code fill:#67E8F9,stroke:#333;
+style manager fill:#FDBA74,stroke:#333;
+```
+
+_The Python memory manager helps manage memory for Python code executed by the Python interpreter._
+
+Memory is managed for Python software processes automatically (when unspecified) or manually (when specified) through the Python interpreter.
+The ___Python memory manager___ is an abstraction which manages memory for Python software processes through the Python interpreter ([Python: Memory Management](https://docs.python.org/3/c-api/memory.html)).
+From a high-level perspective, we assume variables and other operations written in Python will automatically allocate and deallocate memory through the Python interpreter when executed.
+The Python memory manager .
+Python's memory manager performs work through various __memory allocators__ and a __garbage collector__ (or as configured with customizations) within a __private Python memory heap__.
+
+### Python's Memory Allocators
+
+```mermaid!
+flowchart LR
+
+subgraph computer ["Computer"]
+  direction LR
+  subgraph memory["Memory"]
+    pyheap["Python heap"]
+  end
+  code["Python code"]
+  malloc["C memory functions"]
+  subgraph interpreter ["Python interpreter"]
+    subgraph spacer [" "]
+        subgraph mgr ["Python memory manager"]
+            pymalloc["pymalloc"]
+        end
+    end
+  end
+end
+
+code --> |interpreted and\nexecuted by| interpreter
+pymalloc --> |"allocates memory for\nsmall and temporary needs"| malloc
+mgr --> |"allocates memory for\nlarger or long-lived needs"| malloc
+malloc --> pyheap
+
+
+style computer fill:#fff,stroke:#333
+style memory fill:#86EFAC,stroke:#333
+style code fill:#67E8F9,stroke:#333;
+style mgr fill:#FDBA74,stroke:#333;
+style pymalloc fill:#FBBF24,stroke:#333;
+style malloc fill:#FBBF24,stroke:#333;
+style spacer stroke:none;
+```
+
+_The Python memory manager by default will use `pymalloc` internally or malloc from the system to allocate computer memory resources._
+
+The Python memory manager allocates memory for use through memory allocators.
+Python may use one or many memory allocators depending on specifications in Python code and how the Python interpreter is configured (for example, see [Python: Memory Management - Default Memory Allocators](https://docs.python.org/3/c-api/memory.html#default-memory-allocators)).
+One way to understand Python memory allocators is through the following distinctions.
+
+- __"Python Memory Allocator" (`pymalloc`)__
+  The Python interpreter is packaged with a specialized memory allocator called `pymalloc`.
+  "Python has a pymalloc allocator optimized for small objects (smaller or equal to 512 bytes) with a short lifetime." ([Python: Memory Management - The pymalloc allocator](https://docs.python.org/3/c-api/memory.html#the-pymalloc-allocator)).
+  Ultimately, `pymalloc` uses `C malloc` to implement memory work.
```

Comment:
As I wrote this I discovered that `mimalloc` and `jemalloc` make use of duplicate / drop-in function names which are also called `malloc`. I added the `C` here to indicate which `malloc` given that multiple allocators are brought up. I like your change suggestion here, I'll add that in. I find myself wishing there were a way to align the references here (i.e. `cmalloc` or something like that 🙂 ).

Reply:
I see your point, but IMHO those are other allocators are reusing `malloc` because `malloc`, established in the C standard library, is so well-known. If I were to see `malloc` in the wild, I'd assume it was the C `malloc`, even if other libraries reuse the name for compatibility with it.

Up to you. I agree that "C malloc" is the most specific way to state it, so I think it's fine to keep that, although I'd style it "C `malloc`" and not `C malloc`, since the latter looks like an identifier to me and not "C's malloc". I would also support just calling it `malloc` and differentiating the other, later allocators, by their specific names.

(And sorry, I realize this is a real nitpick; feel free to disregard this entirely and just keep your version as-is.)

Reply:
I think you're spot on with your comments, no need to apologize! I moved to more specifically calling out the C standard library based on your suggestions. Generally I believe there's more going on than just C's standard `malloc` (for instance `realloc`, `calloc`, `aligned_alloc`) so I opted to add "functions" instead of just listing `malloc` anymore. I think the same stands for `mimalloc` and `jemalloc`, however these seem to have used the text `malloc` in their names as a shorthand way to tie into the idea of dynamic memory allocation functions without more words.